### DISCUSSION TITLE
数学公式识别如何可视化模型结构

### DISCUSSION THREAD

**Discussion by lm91888** [PaddlePaddle/PaddleOCR#14340] [Q&A]
model.eval()  
with LogWriter(logdir="./log/graph_test/") as writer:
              writer.add_graph(
                  model=model,
                  input_spec=[InputSpec([1, 1, 90, 200])],
                  verbose=True)
报错/home/zlm/.conda/envs/paddle3/lib/python3.10/site-packages/paddle/static/nn/control_flow.py:1540: UserWarning: Return results from different branches in cond are not same type: false_var returned by false_fn is '<class 'paddle.base.libpaddle.pir.Value'>' and true_var of true_fn is '<class 'paddle.base.libpaddle.pir.Value'>'
  warnings.warn(
/home/zlm/.conda/envs/paddle3/lib/python3.10/site-packages/paddle/static/nn/control_flow.py:1540: UserWarning: Return results from different branches in cond are not same type: false_var returned by false_fn is '<class 'paddle.base.libpaddle.pir.Value'>' and true_var of true_fn is '<class 'paddle.base.libpaddle.pir.Value'>'
  warnings.warn(
/home/zlm/.conda/envs/paddle3/lib/python3.10/site-packages/paddle/static/nn/control_flow.py:1540: UserWarning: Return results from different branches in cond are not same type: false_var returned by false_fn is '<class 'paddle.base.libpaddle.pir.Value'>' and true_var of true_fn is '<class 'paddle.base.libpaddle.pir.Value'>'
  warnings.warn(
W1206 11:03:37.938968 240217 pd_api.cc:48892] got different data type, run type promotion automatically, this may cause data type been changed.
/home/zlm/.conda/envs/paddle3/lib/python3.10/site-packages/paddle/utils/inplace_utils.py:61: UserWarning: In static graph mode, masked_fill_() is the same as masked_fill() and does not perform inplace operation.
  warnings.warn(
/home/zlm/.conda/envs/paddle3/lib/python3.10/site-packages/paddle/utils/inplace_utils.py:61: UserWarning: In static graph mode, masked_fill_() is the same as masked_fill() and does not perform inplace operation.
  warnings.warn(
/home/zlm/.conda/envs/paddle3/lib/python3.10/site-packages/paddle/utils/inplace_utils.py:61: UserWarning: In static graph mode, masked_fill_() is the same as masked_fill() and does not perform inplace operation.
  warnings.warn(
/home/zlm/.conda/envs/paddle3/lib/python3.10/site-packages/paddle/static/nn/control_flow.py:1540: UserWarning: Return results from different branches in cond are not same type: false_var returned by false_fn is '<class 'paddle.base.libpaddle.pir.Value'>' and true_var of true_fn is '<class 'bool'>'
  warnings.warn(
/home/zlm/.conda/envs/paddle3/lib/python3.10/site-packages/paddle/static/nn/control_flow.py:1540: UserWarning: Return results from different branches in cond are not same type: false_var returned by false_fn is '<class 'paddle.base.libpaddle.pir.Value'>' and true_var of true_fn is '<class 'paddle.base.libpaddle.pir.Value'>'
  warnings.warn(
Failed to save model graph, error: [Errno 2] No such file or directory: '/tmp/tmpnb5mh2oc/temp.pdmodel'
Traceback (most recent call last):
  File "/home/zlm/.conda/envs/paddle3/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/zlm/.conda/envs/paddle3/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/zlm/.vscode-server/extensions/ms-python.debugpy-2024.12.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/__main__.py", line 71, in <module>
    cli.main()
  File "/home/zlm/.vscode-server/extensions/ms-python.debugpy-2024.12.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 501, in main
    run()
  File "/home/zlm/.vscode-server/extensions/ms-python.debugpy-2024.12.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 351, in run_file
    runpy.run_path(target, run_name="__main__")
  File "/home/zlm/.vscode-server/extensions/ms-python.debugpy-2024.12.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 310, in run_path
    return _run_module_code(code, init_globals, run_name, pkg_name=pkg_name, script_name=fname)
  File "/home/zlm/.vscode-server/extensions/ms-python.debugpy-2024.12.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 127, in _run_module_code
    _run_code(code, mod_globals, init_globals, mod_name, mod_spec, pkg_name, script_name)
  File "/home/zlm/.vscode-server/extensions/ms-python.debugpy-2024.12.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 118, in _run_code
    exec(code, run_globals)
  File "/home/zlm/PaddleOCR/latex/main.py", line 120, in <module>
    main(config, logger, seed)
  File "/home/zlm/PaddleOCR/latex/main.py", line 81, in main
    writer.add_graph(
  File "/home/zlm/.conda/envs/paddle3/lib/python3.10/site-packages/visualdl/writer/writer.py", line 668, in add_graph
    raise e
  File "/home/zlm/.conda/envs/paddle3/lib/python3.10/site-packages/visualdl/writer/writer.py", line 665, in add_graph
    result = translate_graph(model, input_spec, verbose)
  File "/home/zlm/.conda/envs/paddle3/lib/python3.10/site-packages/visualdl/component/graph/exporter.py", line 31, in translate_graph
    model_data = open(os.path.join(tmp, 'temp.pdmodel'), 'rb').read()
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpnb5mh2oc/temp.pdmodel'

**Comment by GreatV** [comment]
根据您的问题和提供的上下文，以下是可能的解决方案和分析：

---

### **1. 关于数学公式模型结构可视化报错问题**

从报错信息来看，问题主要集中在以下几点：
- 使用 `VisualDL` 的 `add_graph` 方法时，尝试保存模型图时出现了 `FileNotFoundError`：`/tmp/tmpnb5mh2oc/temp.pdmodel`。
- 多次出现有关 `paddle.base.libpaddle.pir.Value` 和 `bool` 类型不匹配的警告。
- 提示 `masked_fill_()` 在静态图模式下不支持原地操作。
- 最终由于临时文件路径 `/tmp/tmpnb5mh2oc/temp.pdmodel` 无法找到导致失败。

#### **解决方法：**

1. **检查临时文件路径问题：**
   - `VisualDL` 在保存模型结构时需要临时文件路径。当前路径 `/tmp` 可能没有写权限或被清理。
   - 尝试手动设置一个有写权限的临时目录，例如：
     ```python
     import os
     os.environ['TMPDIR'] = '/path/to/your/temp_dir'
     ```
     确保 `/path/to/your/temp_dir` 是一个有效且有写权限的目录。

2. **输入规格问题：**
   - `input_spec` 定义了输入张量的形状和类型。确保输入张量形状 `[1, 1, 90, 200]` 和模型的实际输入一致。
   - 如果不确定模型的输入形状，可以使用如下方法动态打印模型的输入：
     ```python
     print(model)
     ```
     或者直接尝试使用 Paddle 的 API 检查模型的输入输出。

3. **警告信息的解决：**
   - **类型不匹配警告：**
     警告主要提示 `cond` 分支返回的类型不一致（`paddle.base.libpaddle.pir.Value` vs. `bool`）。通常是由于模型中条件控制流的分支返回值类型不一致导致的。这可能是 Paddle 框架的潜在问题，但它不会直接影响图的保存，暂时可以忽略。
   - **`masked_fill_()` 警告：**
     如果代码中有 `masked_fill_()` 操作，请改为使用 `masked_fill()`，因为静态图模式下不支持原地操作。

4. **VisualDL 日志路径检查：**
   - 确保 `logdir` 路径有效且有写权限。
   - 如果路径不存在，手动创建：
     ```bash
     mkdir -p ./log/graph_test/
     ```

5. **确保代码运行环境：**
   - 使用 Paddle 2.5.0 或更高版本。
   - 确保 `VisualDL` 的版本为最新版本。例如：
     ```bash
     pip install visualdl --upgrade
     ```

---

### **2. 关于导出推理模型报错问题**

您在导出模型时出现路径相关错误（如 `inference.pdmodel` 文件不存在）。以下是可能的原因和解决方案：

#### **问题分析：**
- 导出模型的路径配置错误，导致 `inference.pdmodel` 文件未正确保存到指定目录。
- 提供的 `pretrained_model` 路径不正确，或者模型文件损坏。
- 配置文件 `.yml` 中的某些参数可能与模型或数据不匹配。

#### **解决方案：**

1. **检查模型路径：**
   - 确保 `Global.pretrained_model` 指向的文件路径存在且模型文件有效。例如：
     ```bash
     ls /home/aistudio/output/ch_PP-OCRv4_det_server/best_accuracy.pdparams
     ```
     如果文件不存在，请检查训练输出目录。

   - 确保 `Global.save_inference_dir` 目录有效且有写权限。例如：
     ```bash
     mkdir -p /home/aistudio/Myinfer_model/new_det/
     ```

2. **重试导出：**
   使用以下命令重新导出模型：
   ```bash
   python /home/aistudio/PaddleOCR-release-2.8/tools/export_model.py \
       -c "/home/aistudio/PaddleOCR-release-2.8/configs/det/ch_PP-OCRv4/ch_PP-OCRv4_det_teacher.yml" \
       -o Global.pretrained_model="/home/aistudio/output/ch_PP-OCRv4_det_server/best_accuracy.pdparams" \
       Global.save_inference_dir="/home/aistudio/Myinfer_model/new_det/"
   ```

3. **验证配置文件和参数：**
   - 打开配置文件 `ch_PP-OCRv4_det_teacher.yml`，确保其中的参数（如模型类型、输入大小等）与训练时使用的设置一致。
   - 如果仍然报错，可以尝试使用另一个预训练模型或重新训练。

4. **检查导出后的文件：**
   确保导出后的目录中有以下三个文件：
   - `inference.pdmodel`
   - `inference.pdiparams`
   - `inference.pdiparams.info`

---

### **3. 量化模型转换 ONNX 的问题**

从提供的信息来看，量化模型在转换为 ONNX 格式后，推理时出现类型错误：`tensor(int8)` 不被支持。

#### **问题分析：**
- ONNXRuntime 对某些量化类型（如 `int8`）的支持有限。
- Paddle2ONNX 转换器在处理量化模型时可能未正确处理 `int8` 权重类型。

#### **解决方案：**

1. **尝试使用最新版本的 Paddle2ONNX：**
   确保安装的是最新版本：
   ```bash
   pip install paddle2onnx --upgrade
   ```

2. **检查 ONNX opset 版本：**
   默认情况下，使用 `--opset_version 11`。如果报错，尝试更高的版本（如 `13` 或 `15`）：
   ```bash
   paddle2onnx --model_dir . --model_filename inference.pdmodel \
       --params_filename inference.pdiparams \
       --deploy_backend onnxruntime \
       --save_file ./rec_slim.onnx --opset_version 13 --enable_onnx_checker True
   ```

3. **禁用量化或使用 FP32：**
   如果量化模型在 ONNXRuntime 中存在兼容性问题，可以选择不使用量化模型，改用 FP32 模型：
   ```bash
   paddle2onnx --model_dir . --model_filename inference.pdmodel \
       --params_filename inference.pdiparams \
       --deploy_backend onnxruntime \
       --save_file ./rec_fp32.onnx --opset_version 11 --enable_onnx_checker True
   ```

4. **修复 `tensor(int8)` 问题：**
   如果必须使用量化模型，可以尝试通过 `PaddleSlim` 重新量化并确保权重类型兼容：
   ```python
   paddleslim.quant.quant_post_dynamic(
       model_dir="./ch_PP-OCRv3_rec_infer",
       model_filename="inference.pdmodel",
       params_filename="inference.pdiparams",
       save_model_dir="./output_rec",
       save_model_filename="inference.pdmodel",
       save_params_filename="inference.pdiparams",
       weight_bits=8,
   )
   ```

---

### **4. VisualDL 无法打开日志的问题**

#### **解决方案：**
1. **检查日志文件路径：**
   确保日志文件 `vdlrecords.1727257519.log` 存在且路径正确：
   ```bash
   ls ./runs/mnist_experiment/rec/epoch_200
   ```

2. **升级 VisualDL：**
   ```bash
   pip install visualdl --upgrade
   ```

3. **使用正确的日志路径：**
   启动 VisualDL 时，指定正确的日志路径：
   ```bash
   visualdl --logdir ./runs/mnist_experiment/rec/epoch_200 --host 0.0.0.0 --port 8080
   ```

4. **确保 Paddle 版本兼容：**
   使用 Paddle >= 2.0.0，例如：
   ```bash
   pip install paddlepaddle==2.5.0
   ```

---

### **总结**

通过上述步骤，您应该可以解决数学公式识别模型可视化、推理模型导出和量化模型转换的问题。如果仍有问题，请提供更多具体信息。

---
*Response generated by 🤖 [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*

**Reply by lm91888** [reply]
使用os.environ['TMPDIR'] = '/path/to/your/temp_dir'后还是相同的报错，设置的路径也有权限

**Comment by GreatV** [comment]
估计是visualdl pir支持还没弄好

**Reply by lm91888** [reply]
要等他更新吗

**Reply by GreatV** [reply]
你下一个最新版或者dev版本试试

**Reply by lm91888** [reply]
是用的最新版的，3.0的
